src/app/helpers/useGrudInit.ts

```diff
@@ -0,0 +1,54 @@
+import f from "lodash/fp";
+import { useEffect, useState } from "react";
+import { makeRequest } from "./apiHelper";
+import route from "./apiRoutes";
+import { promisifyAction } from "../redux/redux-helpers";
+import {
+  initLangtags,
+  initAnnotationConfigs
+} from "../constants/TableauxConstants";
+import actions from "../redux/actionCreators";
+import store from "../redux/store";
+import initUserSettings from "./initUserSettings";
+
+export const useGrudInit = () => {
+  const [isInitialized, setInitialized] = useState(false);
+
+  const init = async () => {
+    try {
+      const langtagsRoute = route.toSetting("langtags");
+      const langtagsResponse = await makeRequest({ apiRoute: langtagsRoute });
+      const langtags = langtagsResponse.value;
+
+      initLangtags(langtags);
+
+      const annotRoute = route.toAnnotationConfigs();
+      const annotResponse = await makeRequest({ apiRoute: annotRoute });
+      const annotationConfigs = annotResponse.annotations;
+
+      initAnnotationConfigs(annotationConfigs);
+
+      await promisifyAction(actions.loadTables)();
+      await promisifyAction(actions.queryFrontendServices)();
+
+      store.dispatch(actions.createDisplayValueWorker());
+
+      const settingsResponse = await promisifyAction(actions.getUserSettings)();
+      const settings = settingsResponse.settings;
+
+      if (f.isEmpty(settings)) {
+        await initUserSettings();
+      }
+
+      setInitialized(true);
+    } catch (err) {
+      console.error("Could not init GRUD!", err);
+    }
+  };
+
+  useEffect(() => {
+    void init();
+  }, []);
+
```

Comment:
MMn sollte ein Hook solche Entscheidungen nicht selber treffen dürfen. Das koppelt hier zwei eigentlich getrennte Dinge - das Laden und die Entscheidung, ob geladen werden muss - für den "Enduser" intransparent.

Reply:
Der Hook ist eigtl. nur ein kleines Refactoring, bei dem ich den useState und den useEffect zusammen mit der init- Funktionalität in einen Hook ausgelagert habe.

```
// davor innerhalb von Router.jsx 
const [isInitialized, setInitSuccess] = React.useState(false);

React.useEffect(() => {
  initGrud(setInitSuccess);
}, []);
```

Ich sehe das Problem dahinter nicht, dass das Laden und die Entscheidung, ob geladen werden muss jetzt im Hook gekoppelt sind.

Reply:
Den State da reinzunehmen finde ich gut und richtig. Wie gesagt ist mein Hauptproblem dabei, dass der Caller die Kontrolle verliert.  
Um das wieder zu fixen würde es evtl. reichen, dem Hook ein optionales Dependency-Array mitzugeben, das dann der Hook direkt verwendet, á la 

```
const useGrudInit = (retryOn = []) => {
  ...
  useEffect(.., retryOn);
};
```

Jetzt liegt die Verantwortung wieder beim Caller, und der muss sich nicht per Fire-and-forget darauf verlassen, dass magisch etwas passiert. Theoretisch könnte man auch evtl. einen Fehlerstate noch mit zurückgeben, aber für wirklich notwendig halte ich das nicht.

Reply:
Ok, das kann ich nachvollziehen.
Habe das noch eingebaut.